### PR TITLE
Add optional E2E tests for MCO using feature gate during bootstrap

### PR DESCRIPTION
**Pull request by JoelSpeed**
Related to openshift/machine-config-operator#2668, this PR adds a tech preview IPI step and the relevant configuration for AWS (this is the first 4 commits, a subset of #19845) then adds optional presubmits for the MCO so that in the future, they will have some signal as to whether the clusters bootstrap correctly when a feature gate is added to the manifests as part of the install.

This job should fail until 2668 is merged. If the cluster bootstraps correctly and doesn't degrade on install, then the MCO part of the bootstrap process is working as expected.

CC @yuqi-zhang @rphillips @mrunalp This has come from our conversation on Friday

### CODE REVIEW

**Comment by sinnykumari** [comment]
Don't have full context but wondering if the workflow and job named more generic like e2e-aws-featuregate. That way we can keep it afterward as well when we are post techpreview.

Does it need https://github.com/openshift/machine-config-operator/pull/2668 in order for this e2e to pass?

**Comment by JoelSpeed** [comment]
> Don't have full context but wondering if the workflow and job named more generic like e2e-aws-featuregate

I named it `techpreview` because it applies the `TechPreviewNoUpgrade` feature set, the idea is that it creates the cluster in a tech preview state as a customer may do. I'm happy to name it however the team prefers for the PR status check though if you disagree with my naming. I'd prefer to keep the steps and associated parts named as they are as they form part of a wider `TechPreviewNoUpgrade` testing effort for the release informing jobs

> Does it need openshift/machine-config-operator#2668 in order for this e2e to pass?

Yes it will do. As it's optional, we could merge this early and observe that 2668 fixes it, or we can wait til 2668 merges and see it pass here. Not sure which is the preferred ordering.

I have manually tested the step as part of the other PR and it does cause the clusters to fail on MCO degraded, as we should see from the rehearsal jobs here as well.

**Comment by sinnykumari** [comment]
> > Don't have full context but wondering if the workflow and job named more generic like e2e-aws-featuregate
> 
> I named it `techpreview` because it applies the `TechPreviewNoUpgrade` feature set, the idea is that it creates the cluster in a tech preview state as a customer may do. I'm happy to name it however the team prefers for the PR status check though if you disagree with my naming. I'd prefer to keep the steps and associated parts named as they are as they form part of a wider `TechPreviewNoUpgrade` testing effort for the release informing jobs

thanks for the providing the context. Perhaps, we can name the test `e2e-aws-csi-migration-techpreview`or something similar which can let us know what this test is all about? 

> I have manually tested the step as part of the other PR and it does cause the clusters to fail on MCO degraded, as we should see from the rehearsal jobs here as well.

+1, we can merge this and ensure that this test passes on https://github.com/openshift/machine-config-operator/pull/2668

**Comment by JoelSpeed** [comment]
> Perhaps, we can name the test e2e-aws-csi-migration-techpreviewor something similar which can let us know what this test is all about?

Let's try and gather some suggestions here. I'd be hesitant to name it with `csi-migration` as that is a particular feature, where this will change eventually right? As new releases change the feature gates we wouldn't want to update the test name?
Would `e2e-aws-techpreview-featuregate` or `e2e-aws-techpreview-featuregate-bootstrap` make it more obvious that we are bootstrapping the cluster with a techpreview featuregate?

**Comment by kikisdeliveryservice** [comment]
> Would `e2e-aws-techpreview-featuregate` or `e2e-aws-techpreview-featuregate-bootstrap` make it more obvious that we are bootstrapping the cluster with a techpreview featuregate?

what about e2e-aws-ccm-integration or e2e-aws-cloud-provider-featuregate

**Comment by JoelSpeed** [comment]
> what about e2e-aws-ccm-integration or e2e-aws-cloud-provider-featuregate

These names tie the test suite to a particular set of features, but these features will evolve over time and I don't think the plan is to remove this once all the CSI or CCM feature work is done (I may have misunderstood there).

In code version of below: https://github.com/openshift/release/pull/20372/files#diff-08f6f5bd2d927a18416780ba27bb50a6e31ffd5b0b90b6054ed64cc042de2d64R6-R15

```
apiVersion: config.openshift.io/v1
kind: FeatureGate
metadata:
  annotations:
    include.release.openshift.io/self-managed-high-availability: "true"
    include.release.openshift.io/single-node-developer: "true"
    release.openshift.io/create-only: "true"
  name: cluster
spec:
  featureSet: TechPreviewNoUpgrade
```

This is what we are applying that causes this test to be different to the standard test suite. My hesitancy in calling this CSI or CCM is because as you can see from that, they aren't explicitly called out there. We expect that the features enabled by this `FeatureGate` are changed release to release, the docs will show this if you look at how this doc differs per different versions

- 4.7: https://docs.openshift.com/container-platform/4.7/nodes/clusters/nodes-cluster-enabling-features.html
- 4.5: https://docs.openshift.com/container-platform/4.5/nodes/clusters/nodes-cluster-enabling-features.html

**Comment by sinnykumari** [comment]
I think we are all getting confused here. Goal of renaming e2e test is to keep it meaningful. Any name that justifies what features or set of features this e2e test is all about is fine. For example: suggested name `e2e-aws-techpreview-featuregate` sounds fine to me.

**Comment by JoelSpeed** [comment]
@sinnykumari Agreed, have updated it to `e2e-aws-techpreview-featuregate` as per your comment 🙂

File: ci-operator/jobs/openshift/machine-config-operator/openshift-machine-config-operator-master-presubmits.yaml
Comment (sinnykumari) [inline ci-operator/jobs/openshift/machine-config-operator/openshift-machine-config-operator-master-presubmits.yaml:470]: We need to add `optional: true`

File: ci-operator/jobs/openshift/machine-config-operator/openshift-machine-config-operator-release-4.9-presubmits.yaml
Comment (sinnykumari) [inline ci-operator/jobs/openshift/machine-config-operator/openshift-machine-config-operator-release-4.9-presubmits.yaml:470]: We need to add `optional: true`

File: ci-operator/jobs/openshift/machine-config-operator/openshift-machine-config-operator-master-presubmits.yaml
Comment (JoelSpeed) [reply ci-operator/jobs/openshift/machine-config-operator/openshift-machine-config-operator-master-presubmits.yaml:470]: Oh, I had done that, perhaps the regenerate killed it, will add back!

**Review by sinnykumari — APPROVED** [review]
LGTM.

**Comment by sinnykumari** [comment]
umm, test `ci-operator-config-metadata` failed. Needs fixed, should be mostly ordering or spacing issue

**Comment by sinnykumari** [comment]
Putting hold until `ci-operator-config-metadata`  test is fixed
/hold

**Comment by sinnykumari** [comment]
Some of the file changes made here would also require approval from one the owner from https://github.com/openshift/release/blob/master/ci-operator/step-registry/ipi/OWNERS

**Comment by JoelSpeed** [comment]
/assign @deads2k as you are familiar with the TPNU stuff that we are adding to the step-registry already/ have already approved this in the other PR

@sinnykumari could you re LGTM if you're happy to proceed?

**Comment by sinnykumari** [comment]
I thought my `/approve` earlier would have added lgtm. These bots acts strangely. 
/lgtm

**Comment by deads2k** [comment]
I think this does what it says it does

/approve

**Pull request merged**
